src/cluster.c

```diff
@@ -2218,12 +2225,63 @@ void clusterUpdateSlotsConfigWith(clusterNode *sender, uint64_t senderConfigEpoc
                     newmaster = sender;
                     migrated_our_slots++;
                 }
+
+                /* If the sender who claims this slot is not in the same shard,
+                 * it must be a result of deliberate operator actions. Therefore,
+                 * we should honor it and clear the outstanding migrating_slots_to
+                 * state for the slot. Otherwise, we are looking at a failover within
+                 * the same shard and we should retain the migrating_slots_to state
+                 * for the slot in question */
+                if (server.cluster->migrating_slots_to[j] != NULL) {
+                    if (!areInSameShard(sender, myself)) {
+                        serverLog(LL_WARNING, "Clear migrating destination for slot %d.", j);
+                        server.cluster->migrating_slots_to[j] = NULL;
+                    }
+                }
+
+                /* Handle the case where we are importing this slot and the ownership changes */
+                if (server.cluster->importing_slots_from[j] != NULL) {
```

Comment:
Are we sure people haven't built tooling around this state management anyways? My understanding is that the redis-cli doesn't handle these failure modes todays, so it's unclear why we're trying to preserve them on the server side.

Reply:
Tooling doesn't mean human operators. When resharding is happening, I am assuming people are using tooling to assist with the reshard since it's hard to orchestrate on your own. I want to understand how that behaves with all of these states. More so because we should fix bugs there.

Reply:
> Right this is the goal with this PR.

Are there any such states where automation might take incorrect actions that were previously acceptable. Such as if a migration failed part way through it might have just finalized the states? Or are we assuming that all automation probably wasn't handling these modes.